cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -217,20 +218,20 @@ public static void validateConfig(final SecureStore secureStore,
    * Processes the hashes to a list of given type S
    *
    * @param <T> Type of elements of the output collection
-   * @param <S> Type of elements of the input collection, must implement {@link Supplier<Path>}
+   * @param <S> Type of elements of the input collection, must implement {@link Supplier Supplier&gt;Path&lt;}
    * @param commitMeta Details for the commit including author, committer and commit message
    * @param filesChanged Supplies relative paths to repository root of the updated files
    * @param hashConsumer a {@link BiFunction} that takes a value of the input collection element type (S),
    *                     a filehash and returns a value of the output collection element type (T)
-   * @return a {@link Collection} of elements of type T
+   * @return a pair of commitId and {@link Collection} of elements of type T
    * @throws GitAPIException when the underlying git commands fail
    * @throws NoChangesToPushException when there's no file changes for the commit
    * @throws GitOperationException when failed to get filehash due to IOException or the
    *     {@link PushResult} status is not OK
    * @throws SourceControlException when failed to get the fileHash before push
    */
-  public <T, S extends Supplier<Path>> Collection<T> commitAndPush(CommitMeta commitMeta, Collection<S> filesChanged,
-      BiFunction<S, String, T> hashConsumer)
+  public <T, S extends Supplier<Path>> ImmutablePair<String, Collection<T>> commitAndPush(
```

Comment:
nit: it's usually more readable and extensible to use a new class instead of Pair. Readable since the getters have names instead of first and second. Extensible in case we need to add any other information in the return object in the future.